formula/src/main/java/com/instacart/formula/Effect.kt

```diff
@@ -0,0 +1,47 @@
+package com.instacart.formula
+
+/**
+ * Effect is a function returned within [Transition.Result] which will be executed by
+ * [FormulaRuntime] usually after [Formula] state changes are applied. This function is
+ * the place to call listeners, log events, trigger network requests or database writes,
+ * and etc. Instead of executing effects within [Transition.toResult], deferring execution
+ * allows us to ensure that [Formula] is always in the correct state in case effects
+ * trigger a new event.
+ */
+data class Effect(
+    val type: Type,
+    val executable: () -> Unit,
+) {
+    constructor(executable: () -> Unit): this(Main, executable)
+
+    /**
+     * An effect type which defines the execution timing and threading of the effect. Take
+     * a look at [Unconfined], [Main], [Background].
+     */
+    sealed class Type
+
+    /**
+     * Unconfined effect type indicates that the effect can run on any thread and will be
+     * executed after all state changes are processed.
+     */
+    data object Unconfined : Type()
```

Comment:
nit: should all implementing classes live inside `Type`?

Reply:
I think that the public API is better as such (as type is implied by `Effect.`) 
```kotlin
transition(Effect.Unconfined) { }
```
rather than
```kotlin
transition(Effect.Type.Unconfined) {}
```